pkg/processor/runtime/rpc/connection/socketallocator.go

```diff
@@ -0,0 +1,225 @@
+/*
+Copyright 2024 The Nuclio Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package connection
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"time"
+
+	"github.com/nuclio/nuclio/pkg/common"
+
+	"github.com/nuclio/errors"
+	"github.com/rs/xid"
+)
+
+const (
+	socketPathTemplate = "/tmp/nuclio-rpc-%s.sock"
+	connectionTimeout  = 2 * time.Minute
+)
+
+type SocketAllocator struct {
+	*AbstractConnectionManager
+
+	eventSockets         []*EventSocket
+	controlMessageSocket *ControlMessageSocket
+}
+
+func NewSocketAllocator(abstractConnectionManager *AbstractConnectionManager) *SocketAllocator {
+	return &SocketAllocator{
+		AbstractConnectionManager: abstractConnectionManager,
+		eventSockets:              make([]*EventSocket, 0),
+	}
+}
+
+// Prepare initializes the SocketAllocator by setting up control and event sockets
+// according to the configuration.
+//
+// If SupportControlCommunication is enabled, a control communication socket is created,
+//
+//	wrapped in a ControlMessageSocket, and integrated with the ControlMessageBroker for runtime operations.
+//
+// Creates a minimum number of event sockets (MinConnectionsNum).
+func (sa *SocketAllocator) Prepare() error {
+	if sa.Configuration.SupportControlCommunication {
+		controlConnection, err := sa.createSocketConnection()
+		if err != nil {
+			return errors.Wrap(err, "Failed to create socket connection")
+		}
+		sa.controlMessageSocket = NewControlMessageSocket(
+			sa.Logger,
+			controlConnection,
+			sa.RuntimeConfiguration.ControlMessageBroker)
+	}
+
+	for i := 0; i < sa.MinConnectionsNum; i++ {
+		eventConnection, err := sa.createSocketConnection()
+		if err != nil {
+			return errors.Wrap(err, "Failed to create socket connection")
+		}
+		sa.eventSockets = append(sa.eventSockets, NewEventSocket(sa.Logger,
+			eventConnection, sa))
+	}
+	return nil
+}
+
+func (sa *SocketAllocator) Start() error {
+	if err := sa.startSockets(); err != nil {
+		return errors.Wrap(err, "Failed to start socket allocator")
+	}
+
+	// wait for start if required to
+	if sa.Configuration.WaitForStart {
+		sa.Logger.Debug("Waiting for start")
+		for _, socket := range sa.eventSockets {
+			socket.Start()
+		}
+	}
```

Comment:
Readability-wise, it looks odd that you call `startSockets`, and then when waiting for start you call `socket.Start()`, although you already started them.
I think the abstract method `Start` that simply waits for the channel needs a better name (maybe `waitForStart` too ? )

Reply:
I understand the we do wait, but `socket.Start()` is not starting the socket, it waits for it to start.
I suggested changing the name of `socket.Start()` to `socket.WaitForStart()`

Reply:
@TomerShor renamed 👍